server/core/dependency_injector.py

```diff
@@ -0,0 +1,136 @@
+import inspect
+from collections import ChainMap, defaultdict
+from typing import Dict, List
+
+DependencyGraph = Dict[str, List[str]]
+
+
+class DependencyInjector(object):
+    """
+    Does dependency injection.
+
+    Dependencies are resolved by parameter name. So if a class has init method
+    ```
+    def __init__(self, hello, world):
+        pass
+    ```
+    the injector will look for two dependencies, called `hello` and `world`.
+    These could either be an object registered with `add_injectables`, or an
+    instance of another class passed to `build_classes`.
+
+    Injected arguments are only ever constructed once. So if two classes both
+    depend on an object called `hello`, then they will both receive the same
+    instance of the object called `hello` (whether that is an injectable, or
+    another class in the class list).
+
+    # Example
+    ```
+    class SomeClass(object):
+        def __init__(self, external):
+            self.external = external
+
+    class SomeOtherClass(object):
+        def __init__(self, some_class):
+            self.some_class = some_class
+
+    injector = DependencyInjector()
+    injector.add_injectables(external=object())
+    classes = injector.build_classes({
+        "some_class": SomeClass,
+        "other": SomeOtherClass
+    })
+
+    assert isinstance(classes["some_class"], SomeClass)
+    assert isinstance(classes["other"], SomeOtherClass)
+    assert classes["other"].some_class is classes["some_class"]
+    ```
+
+    """
```

Comment:
What bothers me, is do we need that class? Like yeah, oop is fine, but we don't need it at all in the end.
Another question: do we need to store stuff in local variable? We could return already existing instances.

Reply:
Right, I should add the instance to `injectables` when they're built. That was part of the reason why I put it in a class, so you could call `build_classes` multiple times and inject your classes from the first call.